### PR TITLE
Added network usage metrics

### PR DESCRIPTION
**Pull request by jessberg**
For my networks class project, I am going to compare the "network costs" of centralized and distributed isomorphism.  So I need a way of measuring how much we're sending across the network in the simulator.  This pull request implements that.

### CODE REVIEW

File: libs/sim/simulator.rs [hunk 1]
```diff
@@ -162,14 +164,20 @@ impl<'a> Simulator {
...
         let mut rpc_buffer = vec![];
+        let mut data_over_network_size = 0;
         // tick all elements to generate RPCs
         // this is the send phase. collect all the RPCs
         for (_elem_name, element_obj) in self.elements.iter_mut() {
             let rpcs = element_obj.tick(tick);
             for rpc in &rpcs {
+                if self.record_network_data {
+                    data_over_network_size += rpc.len();
+                }
```
Comment (fruffy) [inline libs/sim/simulator.rs]: You should do this with the rpc buffer instead,  not in the loop and only when the check is enabled.

File: libs/sim/simulator.rs [hunk 2]
```diff
@@ -201,6 +209,14 @@ impl<'a> Simulator {
                 }
             }
         }
+        if self.record_network_data {
+            log::info!(
```
Comment (fruffy) [inline libs/sim/simulator.rs]: I am not sure if you want to collect data using a log. Instead maintain a csv for this that collects the information. Then you can also easily plot it. Although I am not sure what you are planning to use this for.

File: libs/sim/simulator.rs [hunk 3]
```diff
@@ -201,6 +209,14 @@ impl<'a> Simulator {
...
+                "Network use in tick {0} is {1}",
+                tick,
+                data_over_network_size
+            );
+        }
         log::info!("################# TICK {0} END #################", tick);
+        return data_over_network_size;
```
Comment (fruffy) [inline libs/sim/simulator.rs]: I do not like this. tick has a specific purpose and this is tacking on stuff. I would keep all this metrics collection separate. You can compute everything you need in an independent csv

**Review by fruffy — CHANGES_REQUESTED** [review]

File: libs/sim/simulator.rs [hunk 4]
```diff
@@ -166,23 +167,42 @@ impl<'a> Simulator {
...
+                        Err(e) => {
+                            log::error!("Could not write record, {0}", e);
+                        }
+                    }
+                }
+                Err(e) => {
+                    log::error!("Could not open csv writer, {0}", e);
+                }
+            }
+        }
```
Comment (fruffy) [inline libs/sim/simulator.rs:207]: I would make this a function and then this should be okay.

Comment (jessberg) [reply libs/sim/simulator.rs:207]: Sounds good.  I also edited it so that you specify the file name in the arguments.

**Review by fruffy — APPROVED** [review]

**Pull request merged**